benchmark/arm-manipulation-cg.cpp

```diff
@@ -0,0 +1,544 @@
+      boost::make_shared<ADDifferentialActionModelFreeFwdDynamics>(ad_state, ad_actuation, ad_runningCostModel);
+
+  boost::shared_ptr<ADDifferentialActionModelFreeFwdDynamics> ad_terminalDAM =
+      boost::make_shared<ADDifferentialActionModelFreeFwdDynamics>(ad_state, ad_actuation, ad_terminalCostModel);
+
+  // ADVectorXs ad_armature(ad_state->get_nq());
+  // ad_armature << 0.1, 0.1, 0.1, 0.1, 0.1, 0.1, 0.;
+  // ad_runningDAM->set_armature(ad_armature);
+  // ad_terminalDAM->set_armature(ad_armature);
+  boost::shared_ptr<ADActionModelAbstract> ad_runningModel =
+      boost::make_shared<ADIntegratedActionModelEuler>(ad_runningDAM, ADScalar(1e-3));
+  boost::shared_ptr<ADActionModelAbstract> ad_terminalModel =
+      boost::make_shared<ADIntegratedActionModelEuler>(ad_terminalDAM, ADScalar(1e-3));
+
+  /****************************/
+
+  // For calculation and for the ShootingProblem!!
+
+  boost::shared_ptr<crocoddyl::ActionModelAbstractTpl<Scalar> > cg_runningModel =
+      boost::make_shared<crocoddyl::ActionModelCodeGenTpl<Scalar> >(ad_runningModel, runningModel,
+                                                                    "pyrene_model_running");
+  boost::shared_ptr<crocoddyl::ActionModelAbstractTpl<Scalar> > cg_terminalModel =
+      boost::make_shared<crocoddyl::ActionModelCodeGenTpl<Scalar> >(ad_terminalModel, terminalModel,
+                                                                    "pyrene_model_terminal");
+  std::vector<boost::shared_ptr<crocoddyl::ActionModelAbstractTpl<Scalar> > > cg_runningModels(N, cg_runningModel);
+  boost::shared_ptr<crocoddyl::ShootingProblem> cg_problem =
+      boost::make_shared<crocoddyl::ShootingProblem>(x0, cg_runningModels, cg_terminalModel);
+  for (unsigned int i = 0; i < N; ++i) {
+    const boost::shared_ptr<crocoddyl::ActionModelAbstractTpl<Scalar> >& cg_model = cg_problem->get_runningModels()[i];
+    const boost::shared_ptr<crocoddyl::ActionDataAbstractTpl<Scalar> >& cg_data = cg_problem->get_runningDatas()[i];
+    cg_model->quasiStatic(cg_data, us[i], x0);
+  }
+
+  crocoddyl::SolverDDP cg_ddp(cg_problem);
+  cg_ddp.setCandidate(xs, us, false);
+
+  // Check that code-generated action model is the same as original.
+  /**************************************************************************/
+  boost::shared_ptr<crocoddyl::ActionDataAbstractTpl<Scalar> > cg_runningData = cg_runningModel->createData();
+  boost::shared_ptr<crocoddyl::ActionDataAbstractTpl<Scalar> > runningData = runningModel->createData();
+  VectorXs x_rand = cg_runningModel->get_state()->rand();
+  VectorXs u_rand = VectorXs::Random(cg_runningModel->get_nu());
+  runningModel->calc(runningData, x_rand, u_rand);
+  runningModel->calcDiff(runningData, x_rand, u_rand);
+  cg_runningModel->calc(cg_runningData, x_rand, u_rand);
+
+  cg_runningModel->calcDiff(cg_runningData, x_rand, u_rand);
+  assert_pretty(cg_runningData->xnext.isApprox(runningData->xnext), "Problem in xnext");
```

Comment:
It is in the unittest as well.
Rule of thumb: anytime you use a codegen, be sure that you check and confirm the output. pinocchio is not widely tested with codegen, and there are still some sections which are not completely supported.

Reply:
Yes, but this kind of code should go to the unittest. This practice pollutes the code, and makes maintenance process more complex.

Is this test already in the unittest code of this PR?

Reply:
Asserts in benchmarks don't make any sense. The reason to have this here is just to be able to debug the codegen, and to be sure that the code-gen benchmarks are actually producing the required output, and not simply returning the initial zero value. If the benchmarks return the initial value without any computation, it would be super fast, but would not make any sense. I prefer to have asserts to make sure that the code-generated output is actually worth having. Until we have a thorough unit-testing of code-gen, this should be a required practice to follow.

Reply:
I am OK to keep these lines for now.